openmls/src/group/mls_group/proposals.rs

```diff
@@ -205,3 +194,267 @@ impl StagedProposalQueue {
+    /// /// Note that the proposal type will always be `Reference`.
+    pub(crate) fn from_staged_proposal(
+        ciphersuite: &Ciphersuite,
+        backend: &impl OpenMlsCryptoProvider,
+        staged_proposal: &'a StagedProposal,
+    ) -> Result<Self, StagedProposalError> {
+        let proposal_reference =
+            ProposalReference::from_proposal(ciphersuite, backend, staged_proposal.proposal())?;
+        Ok(Self {
+            proposal: staged_proposal.proposal(),
+            proposal_reference,
+            sender: *staged_proposal.sender(),
+            proposal_or_ref_type: ProposalOrRefType::Reference,
+        })
+    }
+    /// Returns the `Proposal` as a reference
+    pub(crate) fn proposal(&self) -> &Proposal {
+        self.proposal
+    }
+    /// Returns the `ProposalReference` as a reference
+    pub(crate) fn proposal_reference(&self) -> ProposalReference {
+        self.proposal_reference.clone()
+    }
+    /// Returns the `Sender` as a reference
+    pub(crate) fn sender(&self) -> &Sender {
+        &self.sender
+    }
+}
+
+/// Proposal queue that helps filtering and sorting the Proposals from one
+/// epoch. The Proposals are stored in a `HashMap` which maps Proposal
+/// references to Proposals, such that, given a reference, a proposal can be
+/// accessed efficiently. To enable iteration over the queue in order, the
+/// `ProposalQueue` also contains a vector of `ProposalReference`s.
+#[derive(Default, Debug)]
+pub struct CreationProposalQueue<'a> {
+    /// `proposal_references` holds references to the proposals in the queue and
+    /// determines the order of the queue.
+    proposal_references: Vec<ProposalReference>,
+    /// `queued_proposals` contains the actual proposals in the queue. They are
+    /// stored in a `HashMap` to allow for efficient access to the proposals.
+    queued_proposals: HashMap<ProposalReference, QueuedProposal<'a>>,
+}
+
+impl<'a> CreationProposalQueue<'a> {
+    /// Filters received proposals
+    ///
+    /// 11.2 Commit
+    /// If there are multiple proposals that apply to the same leaf,
+    /// the committer chooses one and includes only that one in the Commit,
+    /// considering the rest invalid. The committer MUST prefer any Remove
+    /// received, or the most recent Update for the leaf if there are no
+    /// Removes. If there are multiple Add proposals for the same client,
+    /// the committer again chooses one to include and considers the rest
+    /// invalid.
+    ///
+    /// The function performs the following steps:
+    ///
+    /// - Extract Adds and filter for duplicates
+    /// - Build member list with chains: Updates & Removes
+    /// - Check for invalid indexes and drop proposal
+    /// - Check for presence of Removes and delete Updates
+    /// - Only keep the last Update
+    ///
+    /// Return a [CreationProposalQueue] and a bool that indicates whether Updates for the
+    /// own node were included
+    pub(crate) fn filter_proposals(
```

Comment:
I'm not sure how much of this function is copied from the old location, but there are still a few `unwrap`s in here. It would be nice to remove them, but if this is largely just c&p, feel free to leave them in.

Reply:
This is 99& c&p. We have a ticket for all the `unwrap`s and I think it's easier to do it in one go, while implementing a coherent `LibraryError` error type throughout.